src/main/kotlin/se/uulm/snowballr/backend/model/dto/ProjectMember.kt

```diff
@@ -1,18 +1,47 @@
 package se.uulm.snowballr.backend.model.dto
 
+import se.uulm.snowballr.backend.model.SnowballRException
 import se.uulm.snowballr.backend.table.association.ProjectMemberTable
 import snowballr.ProjectOuterClass
+import snowballr.UserOuterClass
 import java.time.OffsetDateTime
 import java.util.UUID
+import snowballr.ProjectOuterClass.Project.Member.List as MemberList
 
 /**
  * DTO of [ProjectMemberTable].
  */
 data class ProjectMember(
-    val id: String,
     val projectId: UUID,
     val userId: UUID,
     val role: ProjectOuterClass.MemberRole,
     val createdAt: OffsetDateTime,
     val modifiedAt: OffsetDateTime?,
 )
+
+/**
+ * Creates a [ProjectOuterClass.Project.Member] from this [ProjectMember].
+ */
+fun ProjectMember.toGrpcProjectMember(user: UserOuterClass.User): ProjectOuterClass.Project.Member {
+    return ProjectOuterClass.Project.Member
+        .newBuilder()
+        .setRole(role)
+        .setUser(user)
+        .build()
+}
+
+/**
+ * Creates a list of [ProjectOuterClass.Project.Member]s from this list of [ProjectMember]s.
+ */
+@Suppress("UnreachableCode")
+fun List<ProjectMember>.toGrpcProjectMembers(users: List<User>): MemberList {
+    val userMap = users.associateBy { it.id } // UUID → User
+
+    val builder = MemberList.newBuilder()
+    this.forEach { member ->
+        val user = userMap[member.userId]
+            ?: throw SnowballRException.GrpcConversionException.GrpcList(member.userId, MemberList::class)
+        builder.addMembers(member.toGrpcProjectMember(user.toGrpcUser()))
+    }
+    return builder.build()
+}
```

Comment:
I think your implementation is already excellent. I only would remove the exception. Since the userId is a reference to another table column with restraints, it is by design not possible that a project member exists but no corresponding user.

```suggestion
fun List<ProjectMember>.toGrpcProjectMembers(users: List<User>): MemberList {
    val userMap = users.associateBy { it.id }

    val projectMembers = this.map { member ->
        val user = userMap[member.userId]
        user?.let {
            member.toGrpcProjectMember(it.toGrpcUser())
        }
    }

    val builder = MemberList.newBuilder()
    builder.addAllMembers(projectMembers)
    return builder.build()
}
```

Reply:
All right. Thank you for the suggestion. I was also thinking about this. But in fact you are right and we don't need the exception.